### PR TITLE
ByteBuffer: small & medium sized string copies over 10% faster

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

urbp[start ..< end] is quite a bit faster than when done in two
operations urbp.dropFirst(start).prefix(length). This improves small (4
bytes) & medium sized (24 bytes) String copies by over 10%. Good for
HTTP headers etc.

Modifications:

replace urbp.dropFirst(start).prefix(length) by urbp[start ..< start +
length]

Result:

small & medium sized String copies faster

### CODE REVIEW

**Comment by weissi** [comment]
spurious test error

    Fatal error: Posix wrapper adds more than 20% overhead (with wrapper: 0.72994601726532, without: 0.600885033607483

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Sources/NIO/ByteBuffer-core.swift [hunk 1]
```diff
@@ -421,9 +421,7 @@ public struct ByteBuffer {
                "called the slower set<S: Sequence> function even though \(S.self) is a ContiguousCollection")
         func ensureCapacityAndReturnStorageBase(capacity: Int) -> UnsafeMutablePointer<UInt8> {
             self._ensureAvailableCapacity(_Capacity(capacity), at: index)
-            let newBytesPtr = UnsafeMutableRawBufferPointer(rebasing: self._slicedStorageBuffer
-                .dropFirst(Int(index))
-                .prefix(capacity))
+            let newBytesPtr = UnsafeMutableRawBufferPointer(rebasing: self._slicedStorageBuffer[Int(index) ..< Int(index) &+ Int(capacity)])
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/ByteBuffer-core.swift]: @Lukasa just to double check I understand you are talking about `&+` (no overflow check) ?

[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIO/ByteBuffer-core.swift]: Correct

[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/ByteBuffer-core.swift]: I think I am generally fine with doing it here but if we do so we should also do it in other places where we are sure overflow can never happen to be consist. So maybe we should just not do it in this PR and then open another PR with only the "unchecked" stuff.

[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/ByteBuffer-core.swift]: Sounds good to me!

[the remark this answers is not in the document]
Comment (weissi) [reply Sources/NIO/ByteBuffer-core.swift]: benchmarks don't show that the `&+` make a diff here, removing

**Pull request merged**